x/wasm/watcher/write.go

```diff
@@ -0,0 +1,80 @@
+package watcher
+
+import (
+	"log"
+
+	sdk "github.com/okex/exchain/libs/cosmos-sdk/types"
+)
+
+func NewHeight() {
+	if Enable() && len(blockStateCache) != 0 {
+		blockStateCache = make(map[string]*WatchMessage)
+	}
+}
+
+func Save(err error) {
+	if !Enable() {
+		return
+	}
+	txCacheMtx.Lock()
+	if err == nil {
+		for _, msg := range txStateCache {
+			blockStateCache[string(msg.Key)] = msg
+		}
+	}
+	txStateCache = txStateCache[:0]
+	txCacheMtx.Unlock()
+}
+
+func Commit() {
+	if !Enable() {
+		return
+	}
+	blockStateCacheCopy := blockStateCache
+	task := func() {
+		batch := db.NewBatch()
+		for _, msg := range blockStateCacheCopy {
+			if msg.IsDelete {
+				batch.Delete(msg.Key)
+			} else {
+				batch.Set(msg.Key, msg.Value)
+			}
+		}
+		if err := batch.Write(); err != nil {
+			log.Println("wasm watchDB batch write error:", err.Error())
+		}
+	}
+	tasks <- task
+}
+
+var tasks = make(chan func(), 5*3)
+
+func taskRoutine() {
```

Comment:
if node graceful shutdown. this will cause losing latest version watchdb data.

Reply:
it will be fixed in a new PR